storage_s3/indico_storage_s3/plugin.py

```diff
@@ -0,0 +1,148 @@
+#
+# Indico is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License as
+# published by the Free Software Foundation; either version 3 of the
+# License, or (at your option) any later version.
+#
+# Indico is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Indico; if not, see <http://www.gnu.org/licenses/>.
+
+from __future__ import unicode_literals
+
+import sys
+from contextlib import contextmanager
+from datetime import datetime
+from tempfile import NamedTemporaryFile
+
+import boto3
+import botocore
+from werkzeug.datastructures import Headers
+from werkzeug.utils import redirect
+
+from indico.core import signals
+from indico.core.config import config
+from indico.core.plugins import IndicoPlugin
+from indico.core.storage import Storage, StorageError
+
+
+class S3StoragePlugin(IndicoPlugin):
+    """S3 Storage
+
+    Provides S3 storage backends.
+    """
+
+    def init(self):
+        super(S3StoragePlugin, self).init()
+        self.connect(signals.get_storage_backends, self._get_storage_backends)
+
+    def _get_storage_backends(self, sender, **kwargs):
+        return S3Storage
+
+
+class S3Storage(Storage):
+    name = 's3'
+    simple_data = False
+
+    def __init__(self, data):
+        data = self._parse_data(data)
+        self.host = data['host']
+        self.bucket = data['bucket']
+        self.secret_key = data['secret_key']
+        self.access_key = data['access_key']
+        self.acl_template_bucket = data['acl_template_bucket'] if 'acl_template_bucket' in data else None
+        self.client = boto3.client(
+            's3',
+            endpoint_url=self.host,
+            aws_access_key_id=self.access_key,
+            aws_secret_access_key=self.secret_key,
+        )
+
+    def open(self, file_id):
+        try:
+            return self.client.get_object(Bucket=self._get_current_bucket(), Key=file_id)['Body']
+        except Exception as e:
+            raise StorageError('Could not open "{}": {}'.format(file_id, e)), None, sys.exc_info()[2]
+
+    @contextmanager
+    def get_local_path(self, file_id):
+        with NamedTemporaryFile(suffix='indico.tmp', dir=config.TEMP_DIR) as tmpfile:
+            self._copy_file(self.open(file_id), tmpfile)
+            tmpfile.flush()
+            yield tmpfile.name
+
+    def save(self, name, content_type, filename, fileobj):
+        try:
+            fileobject = self._ensure_fileobj(fileobj)
+            bucket = self._get_current_bucket()
+            self.client.upload_fileobj(fileobject, bucket, name)
+            checksum = self.client.head_object(Bucket=bucket, Key=name)['ETag'][1:-1]
```

Comment:
Do we know for sure that this is an md5 checksum (both ceph and s3)? Also out of curiosity, what are the first/last chars you are stripping off?